agent/processwatcher.go

```diff
@@ -0,0 +1,434 @@
+		buffer.WriteString(fmt.Sprintf("PID: %d", proc.Pid))
+
+		if proc.Outdated {
+			buffer.WriteString(", is running outdated lib(s)")
+		}
+
+		buffer.WriteString(fmt.Sprintf("\nCommand: %s", proc.Command))
+
+		for _, lib := range proc.Libraries {
+			buffer.WriteString("\n")
+			libraryToString(buffer, lib.Outdated, pm.libraries[lib.libraryIndex])
+		}
+	}
+	return buffer.String()
+}
+
+func libraryToString(buffer *bytes.Buffer, outdated bool, lib libspector.Library) {
+	if outdated {
+		buffer.WriteString("Outdated: yes")
+	} else {
+		buffer.WriteString("Outdated:  no")
+	}
+
+	buffer.WriteString(fmt.Sprintf(", Path: %v, ", lib.Path()))
+
+	if pkg, err := lib.Package(); err != nil {
+		log.Warningf("Error reading package info for %s: %v", lib.Path(), err)
+		buffer.WriteString("Package: unknown, ")
+	} else {
+		buffer.WriteString(fmt.Sprintf("Package: %s-%s, ", pkg.Name(), pkg.Version()))
+	}
+
+	if modified, err := lib.Modified(); err != nil {
+		log.Warningf("Error reading modification date: %v", err)
+		buffer.WriteString("Modified: unknown")
+	} else {
+		buffer.WriteString(fmt.Sprintf("Modified: %v", modified))
+	}
+
+	return
+}
+
+func (pm *processMap) MarshalJSON() ([]byte, error) {
+	libraries := make([]map[string]interface{}, len(pm.libraries))
+	for i, lib := range pm.libraries {
+		libraries[i] = libToMap(lib)
+	}
+
+	processes := make([]map[string]interface{}, len(pm.processes))
+	for i, proc := range pm.processes {
+		processes[i] = map[string]interface{}{
+			"started":   proc.ProcessStartedAt,
+			"libraries": procLibsToMapArray(proc.Libraries),
+			"outdated":  proc.Outdated,
+			"pid":       proc.Pid,
+			"name":      proc.Command,
+		}
+	}
+
+	return json.Marshal(map[string]interface{}{
+		"processes": processes,
+		"libraries": libraries,
+	})
+}
+
+func procLibsToMapArray(libs []processLibrary) []map[string]interface{} {
+	procLibs := make([]map[string]interface{}, len(libs))
+
+	for i, lib := range libs {
+		procLibs[i] = map[string]interface{}{
+			"outdated":      lib.Outdated,
+			"library_index": lib.libraryIndex,
+		}
+	}
+
+	return procLibs
+}
+
+func libToMap(lib libspector.Library) map[string]interface{} {
+	path := lib.Path()
+
+	modified, err := lib.Modified()
+	if err != nil {
+		log.Warningf("error retrieving modification date for lib %s, %v", path, err)
+		return nil
+	}
+
+	pkgName := ""
+	pkgVersion := ""
+	if pkg, err := lib.Package(); err != nil {
+		log.Warningf("error retrieving package name for lib %s, %v", path, err)
+	} else {
+		pkgName = pkg.Name()
+		pkgVersion = pkg.Version()
+	}
+
+	return map[string]interface{}{
+		"path":            path,
+		"modified":        modified,
+		"package_name":    pkgName,
+		"package_version": pkgVersion,
+	}
+}
+
+func (pm *processMap) findLibraryIndex(path string) int {
+	for i, library := range pm.libraries {
+		if library.Path() == path {
+			return i
+		}
+	}
+
+	return -1
+}
+
+func (pm *processMap) maybeAddLibrary(lib libspector.Library) int {
```

Comment:
"maybe" here denotes that the func does a set operation - won't add it if it already exists.

I guess I expect that to mean it could sometimes work, sometimes not work :P. Perhaps just "addLibrary"?